src/doc/reference/reducers.md

```diff
@@ -0,0 +1,257 @@
+_hyperobject_.  Different types of hyperobjects are used depending on
+the desired semantics.
+
+Reducers are used when the final value in a variable is built up from
+a series of independent modifications, such as adding a number to an
+accumulator or appending an item to a list.  As long as the operation
+is _associative_ (`A ⊕ (B ⊕ C) = (A ⊕ B) ⊕ C`) the final result will
+be correct.
+
+Formally, a reducer is a mathematical object called a _{% defn
+"monoid" %}_.  A reducer has a type (e.g., `double`), an _identity_
+value (`0.0`), and an associative binary operation (`+`).  The
+operation does not need to be commutative.  A reducer can hold a list
+with the binary operation being concatenation.  Associativity is
+essential.
+
+## Reducers and views
+
+OpenCilk ensures that every reference to a reducer uses a private
+copy, called a _view_.  The address of the current view can change at
+any spawn or sync, including the implicit spawns and syncs associated
+with `cilk_for` and `cilk_scope`.  The address operator `&` returns
+the address of the current view, so the address of a reducer can
+change when the address of a normal variable would be constant over
+its lifetime.  Be careful about saving the address of a reducer.  The
+race detector (Cilk sanitizer) can be used to check for improper
+retention of a pointer to a view.
+
+Views are created and merged using programmer-provided callback
+functions commonly named `identity` and `reduce`.  The identity
+callback takes a pointer to the value to be initialized (cast to
+`void`&nbsp;`*`).  The reduce callback takes two pointer arguments
+pointing to the two values to be combined.  The value pointed to by
+the second argument should be merged into the value pointed to by the
+first argument, and storage held by the second argument should be
+freed.  Even if the operation is commutative, the result should be
+stored in the first argument.
+
+There is a total order on views, the order in which they would have
+been created in a {% defn "serial projection", "serial" %} program.
+The older of any pair of views is conventionally called the _left_
+view and the younger of the pair is called the _right_ view.  The left
+view is the first argument to the reduce callback.  The variable
+declared by the programmer is the _leftmost_ view.  The programmer
+needs to initialize or construct the variable just like any other.
+See `<cilk/ostream_reducer.h>` for an example where the leftmost view
+does not get the identity value.
+
+## Declaring a reducer
+
+A reducer is declared with the `cilk_reducer` keyword, with the
+identity and reduce functions as arguments.
+
+For example, to declare a reducer holding sums of `double`s
+one can write
+
+```c
+    void zero_double(void *view) { *(double *)view = 0.0; }
+    void add_double(void *left, void *right)
+        { *(double *)left += *(double *)right; }
+    double cilk_reducer(zero_double, add_double) sum;
+```
```

Comment:
My inclination is to remove the leading indentation on the code within the code block (for this code block and all other code blocks in this document).  Right now this indentation appears to add unnecessary whitespace to the code on the deployed page.